automatic/nexus-repository/tools/helpers.ps1

```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
+
+    if ((Test-Path $KeyToolPath) -and (Test-Path $KeyStorePath)) {
+      # Running in a job, as otherwise KeyTool fails when run without input
+      Start-Job {
+        $KeyToolOutput = $using:KeyStorePassword | & "$using:KeyToolPath" -list -v -keystore "$using:KeyStorePath" -J"-Duser.language=en" 2>$null
+        if ($KeyToolOutput -join "`n" -match "(?smi)Certificate\[1\]:\nOwner: CN=(?<Domain>.+?)\n") {
+          $Matches.Domain
+        }
+      } | Receive-Job -Wait
+    }
   }
 }
 
-function Wait-NexusAvailability {
+function Get-NexusUri {
   param(
-    [Parameter(Mandatory = $true)]
-    [string]$Hostname,
+    [Parameter()]
+    [string]$DataDir = (Join-Path $env:ProgramData "sonatype-work"),
 
-    [Parameter(Mandatory = $true)]
-    [uint16]$Port,
+    [Parameter()]
+    [string]$ProgramDir = (Join-Path $env:ProgramData "nexus"),
 
-    [Parameter(Mandatory = $true)]
-    [Alias("Config")]
-    [string]$NexusConfigFile,
+    [Parameter()]
+    [string]$ConfigPath = $(
+      if (Test-Path $DataDir/nexus3/etc/nexus.properties) {
+        "$DataDir/nexus3/etc/nexus.properties"
+      } elseif (Test-Path $ProgramDir/etc/nexus-default.properties) {
+        "$ProgramDir/etc/nexus-default.properties"
+      }
+    ),
 
-    [switch]$SSL
+    [string]$HostnameOverride
   )
-  # Even though windows reports service is ready - web url will not respond until Nexus is actually ready to serve content
-  # We need to use this method to collect the port number so we can properly test the website has returned OK.
-  $nexusScheme, $portConfigLine = if ($SSL) {
-    # This is to combat Package Internalizer's over-enthusiastic URL matching
-        ('http' + 's'), 'application-port-ssl'
-  } else {
-    'http', 'application-port'
+  $Scheme, $Hostname, $Port, $Path = if (Test-Path $ConfigPath) {
+    $Config = Get-NexusConfiguration -Path $ConfigPath -ErrorAction SilentlyContinue
+
+    if ($Config.'application-port-ssl' -gt 0) {
+      # This is to combat Package Internalizer's over-enthusiastic URL matching
+      ('http' + 's')
+      if ($CertDomain = Get-NexusCertificateDomain -ConfigPath $ConfigPath) {
+      if (-not $script:OverriddenDomains) {$script:OverriddenDomains = @{}}
+        if ($CertDomain -notmatch '^\*') {
+          $CertDomain
+        } elseif ($CertDomain -match '^\*' -and $HostnameOverride -like $CertDomain) {
+          ($script:OverriddenDomains[$CertDomain] = $HostnameOverride)
+        } elseif ($CertDomain -match '^\*') {
+          while ($script:OverriddenDomains[$CertDomain] -notlike $CertDomain) {
+            $script:OverriddenDomains[$CertDomain] = Read-Host "Please provide the FQDN for Nexus matching the '$($CertDomain)' certificate"
+          }
+          $script:OverriddenDomains[$CertDomain]
+        }
+      } else {
+        Write-Warning "Could not figure out SSL configuration for $($env:ComputerName)"
+        $env:ComputerName
```

Comment:
I would _personally_ prefer this to default to localhost rather than computer name. Some environments (mine 😅) aren't going to resolve the computer name to something useful for the purposes of ensuring Nexus started.